src/item.cpp

```diff
@@ -13047,151 +13111,169 @@ bool item::process_cable( map &here, Character *carrier, const tripoint &pos, it
+int item::charge_linked_batteries( vehicle &linked_veh, int turns_elapsed )
 {
-    if( !link || link->charge_rate == 0 || turns_elapsed < 1 || link->charge_interval < 1 ) {
-        return 0;
+    if( link->charge_rate == 0 || turns_elapsed < 1 ||
+        link->charge_interval < 1 || link->efficiency < 0.001f ) {
+        return link->charge_rate;
     }
 
-    if( !linked_item.is_battery() ) {
-        const item *parent_mag = linked_item.magazine_current();
+    if( !is_battery() ) {
+        const item *parent_mag = magazine_current();
         if( !parent_mag || !parent_mag->has_flag( flag_RECHARGE ) ) {
             return 0;
         }
     }
 
     const bool power_in = link->charge_rate > 0;
-    if( power_in ? linked_item.ammo_remaining() >= linked_item.ammo_capacity( ammo_battery ) :
-        linked_item.ammo_remaining() <= 0 ) {
+    if( power_in ? ammo_remaining() >= ammo_capacity( ammo_battery ) :
+        ammo_remaining() <= 0 ) {
         return 0;
     }
 
-    // Normally efficiency is a random chance to skip a charge, but if we're catching up from time
-    // spent ouside the reality bubble it should be applied as a percentage of the total instead.
+    // Normally efficiency is the chance to get a charge every charge_interval, but if we're catching up from
+    // time spent ouside the reality bubble it should be applied as a percentage of the total instead.
     bool short_time_passed = turns_elapsed <= link->charge_interval;
 
-    if( !calendar::once_every( time_duration::from_turns( link->charge_interval ) ) &&
-        short_time_passed ) {
+    if( short_time_passed &&
+        !calendar::once_every( time_duration::from_turns( link->charge_interval ) ) ) {
         return link->charge_rate;
     }
 
     // If a long time passed, multiply the total by the efficiency rather than cancelling a charge.
     int transfer_total = short_time_passed ? 1 :
-                         ( turns_elapsed * 1.0f / link->charge_interval ) * ( 1.0 - 1.0 / link->charge_efficiency );
+                         ( turns_elapsed * 1.0f / link->charge_interval ) * link->charge_interval;
 
     if( power_in ) {
         const int battery_deficit = linked_veh.discharge_battery( transfer_total, true );
         // Around 85% efficient by default; a few of the discharges don't actually recharge
-        if( battery_deficit == 0 && !( short_time_passed && one_in( link->charge_efficiency ) ) ) {
-            linked_item.ammo_set( itype_battery, linked_item.ammo_remaining() + transfer_total );
+        if( battery_deficit == 0 && !( short_time_passed && rng_float( 0.0, 1.0 ) > link->efficiency ) ) {
+            ammo_set( itype_battery, ammo_remaining() + transfer_total );
         }
     } else {
         // Around 85% efficient by default; a few of the discharges don't actually charge
-        if( !( short_time_passed && one_in( link->charge_efficiency ) ) ) {
+        if( !( short_time_passed && rng_float( 0.0, 1.0 ) > link->efficiency ) ) {
```

Comment:
```suggestion
        if(  !short_time_passed || rng_float( 0.0, 1.0 ) <= link->efficiency  ) {
```
Reference: https://clang.llvm.org/extra/clang-tidy/checks/readability/simplify-boolean-expr.html